// rc-menu/lib/SubMenu.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var React = require('react');
var rcUtil = require('rc-util');
var joinClasses = rcUtil.joinClasses;
var classSet = rcUtil.classSet;
var guid = rcUtil.guid;
var KeyCode = rcUtil.KeyCode;
var Menu = require('./Menu');
var createChainedFunction = rcUtil.createChainedFunction;

var SubMenu = React.createClass({
  displayName: 'SubMenu',

  propTypes: {
    openOnHover: React.PropTypes.bool,
    title: React.PropTypes.node,
    onClick: React.PropTypes.func
  },

  mixins: [require('./SubMenuStateMixin')],

  getInitialState: function getInitialState() {
    return {
      activeFirst: false
    };
  },

  saveMenuInstance: function saveMenuInstance(c) {
    this.menuInstance = c;
  },

  _getPrefixCls: function _getPrefixCls() {
    return this.props.rootPrefixCls + '-submenu';
  },

  _getActiveClassName: function _getActiveClassName() {
    return this._getPrefixCls() + '-active';
  },

  _getDisabledClassName: function _getDisabledClassName() {
    return this._getPrefixCls() + '-disabled';
  },

  componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
    if (!nextProps.active) {
      this.setOpenState(false);
    }
  },

  getDefaultProps: function getDefaultProps() {
    return {
      openOnHover: true,
      onMouseEnter: function onMouseEnter() {},
      title: ''
    };
  },

  handleKeyDown: function handleKeyDown(e) {
    var keyCode = e.keyCode;
    var menu = this.menuInstance;

    if (keyCode === KeyCode.ENTER) {
      this.handleClick(e);
      this.setState({
        activeFirst: true
      });
      return true;
    }

    if (keyCode === KeyCode.RIGHT) {
      if (this.state.open) {
        menu.handleKeyDown(e);
      } else {
        this.setOpenState(true);
        this.setState({
          activeFirst: true
        });
      }
      return true;
    }
    if (keyCode === KeyCode.LEFT) {
      var handled;
      if (this.state.open) {
        handled = menu.handleKeyDown(e);
      } else {
        return undefined;
      }
      if (!handled) {
        this.setOpenState(false);
        handled = true;
      }
      return handled;
    }

    if (this.state.open && (keyCode === KeyCode.UP || keyCode === KeyCode.DOWN)) {
      return menu.handleKeyDown(e);
    }
  },

  handleMouseEnter: function handleMouseEnter() {
    var props = this.props;
    props.onHover(props.eventKey);
    if (props.openOnHover) {
      this.setOpenState(true);
      this.setState({
        activeFirst: false
      });
    }
  },

  handleMouseLeave: function handleMouseLeave() {
    if (!this.state.open) {
      this.props.onHover(null);
    }
  },

  handleClick: function handleClick() {
    this.setOpenState(true);
    this.setState({
      activeFirst: false
    });
  },

  handleSubMenuClick: function handleSubMenuClick(key, menuItem, e) {
    this.props.onClick(key, menuItem, e);
  },

  handleSelect: function handleSelect(childKey, child, e) {
    // propagate
    this.props.onSelect(childKey, child, e);
  },

  handleDeselect: function handleDeselect() {
    this.props.onDeselect.apply(null, arguments);
  },

  render: function render() {
    var props = this.props;
    var classes = {};
    var prefixCls = this._getPrefixCls();
    classes[this._getOpenClassName()] = this.state.open;
    classes[this._getActiveClassName()] = props.active;
    classes[this._getDisabledClassName()] = props.disabled;
    this._menuId = this._menuId || guid();
    classes[prefixCls] = true;
    var clickEvents = {};
    var mouseEvents = {};
    var titleMouseEvents = {};
    if (!props.disabled) {
      clickEvents = {
        onClick: this.handleClick
      };
      mouseEvents = {
        onMouseLeave: this.handleMouseLeave
      };
      // only works in title, not outer li
      titleMouseEvents = {
        onMouseEnter: this.handleMouseEnter
      };
    }
    return React.createElement(
      'li',
      _extends({ className: joinClasses(props.className, classSet(classes)) }, mouseEvents),
      React.createElement(
        'div',
        _extends({
          className: prefixCls + '-title'
        }, titleMouseEvents, clickEvents, {
          'aria-expanded': props.active,
          'aria-owns': this._menuId,
          'aria-haspopup': 'true'
        }),
        props.title
      ),
      this.renderChildren(props.children)
    );
  },
  renderChildren: function renderChildren(children) {
    if (!this.state.open) {
      // prevent destroy
      return this._cacheMenu || null;
    }
    var childrenCount = React.Children.count(children);
    var baseProps = {
      sub: true,
      focusable: false,
      onClick: this.handleSubMenuClick,
      onSelect: this.handleSelect,
      onDeselect: this.handleDeselect,
      activeFirst: this.state.activeFirst,
      multiple: this.props.multiple,
      id: this._menuId,
      ref: this.saveMenuInstance
    };
    if (childrenCount === 1 && children.type === Menu) {
      var menu = children;
      baseProps.ref = createChainedFunction(menu.ref, this.saveMenuInstance);
      baseProps.onClick = createChainedFunction(menu.props.onClick, this.handleSubMenuClick);
      this._cacheMenu = React.cloneElement(menu, baseProps);
    } else {
      this._cacheMenu = React.createElement(
        Menu,
        baseProps,
        children
      );
    }
    return this._cacheMenu;
  }
});

module.exports = SubMenu;
